pageserver/src/storage_sync.rs

```diff
@@ -129,1441 +114,660 @@
-                            );
-
-                            register_sync_status(sync_id, sync_start, TASK_NAME, Some(true));
-                            return DownloadStatus::Downloaded;
-                        }
-                        Err(e) => {
-                            error!("Timeline {sync_id} was expected to be in the remote index after a successful download, but it's absent: {e:?}");
-                        }
-                    };
-                }
-                Err(e) => {
-                    error!("Failed to update local timeline metadata: {e:?}");
-                    download_data.retries += 1;
-                    sync_queue.push(sync_id, SyncTask::Download(download_data));
-                    register_sync_status(sync_id, sync_start, TASK_NAME, Some(false));
-                }
-            }
-        }
+        info!("scheduled layer file upload {}", path.display());
+
+        // Launch the task immediately, if possible
+        self.launch_queued_tasks(upload_queue);
+        Ok(())
     }
 
-    DownloadStatus::Nothing
-}
+    ///
+    /// Launch a delete operation in the background.
+    ///
+    /// The deletion won't actually be performed, until all preceding
+    /// upload operations have completed succesfully.
+    pub fn schedule_layer_file_deletion(self: &Arc<Self>, paths: &[PathBuf]) -> anyhow::Result<()> {
+        let mut upload_queue = self.upload_queue.lock().unwrap();
+        ensure!(
+            upload_queue.latest_metadata.is_some(),
+            "upload queue not initialized"
+        );
 
-async fn update_local_metadata(
-    conf: &'static PageServerConf,
-    sync_id: TenantTimelineId,
-    remote_timeline: Option<&RemoteTimeline>,
-) -> anyhow::Result<()> {
-    let remote_metadata = match remote_timeline {
-        Some(timeline) => &timeline.metadata,
-        None => {
-            debug!("No remote timeline to update local metadata from, skipping the update");
-            return Ok(());
+        // Update the remote index file, removing the to-be-deleted files from the index,
+        // before deleting the actual files.
+        for path in paths {
+            let relative_path = RelativePath::from_local_path(
+                &self.conf.timeline_path(&self.timeline_id, &self.tenant_id),
+                path,
+            )?;
+            upload_queue.latest_files.remove(&relative_path);
+        }
+        let disk_consistent_lsn = upload_queue
+            .latest_metadata
+            .as_ref()
+            .unwrap()
+            .disk_consistent_lsn();
+        let index_part = IndexPart::new(
+            upload_queue.latest_files.clone(),
+            disk_consistent_lsn,
+            upload_queue.latest_metadata.as_ref().unwrap().to_bytes()?,
+        );
+        upload_queue
+            .queued_operations
+            .push_back(UploadOp::UploadMetadata(index_part, disk_consistent_lsn));
+
+        // schedule the actual deletions
```

Comment:
If we crash after index_part.json upload finishes but before deletions finish, we'll leak the layer files on S3.

Is this a regression compared to what we had before?
If so, we need to fix it.
Otherwise, we need to create an issue for this and link it with an `XXX` comment from here.

Reply:
It's a known issue. We just accept it for now.

Similarly after layer flush, if we upload some layer files, and crash before the index file has been uploaded, we leak the new files in S3.

Reply:
Ok